src/templates/en/2019/methodology.html

```diff
@@ -310,7 +310,7 @@ <h3 id="wappalyzer"><a href="#wappalyzer" class="anchor-link">Wappalyzer</a></h3
       </p>
 
       <p>
-        Wappalyzer powers many chapters that analyze the popularity of developer tools like WordPress, Bootstrap, and jQuery. For example, the <a href="{{ url_for('chapter', year=year, lang=lang, chapter='ecommerce') }}">Ecommerce</a> and <a href="{{ url_for('chapter', year=year, lang=lang, chapter='cms') }}">CMS</a> chapters rely heavily on the respective <a href="https://www.wappalyzer.com/categories/ecommerce">Ecommerce</a> and <a href="https://www.wappalyzer.com/categories/cms">CMS</a> categories of technologies detected by Wappalyzer.
+        Wappalyzer powers many chapters that analyze the popularity of developer tools like WordPress, Bootstrap, and jQuery. For example, the <a href="./ecommerce">Ecommerce</a> and <a href="./cms">CMS</a> chapters rely heavily on the respective <a href="https://www.wappalyzer.com/categories/ecommerce">Ecommerce</a> and <a href="https://www.wappalyzer.com/categories/cms">CMS</a> categories of technologies detected by Wappalyzer.
```

Comment:
Why is `url_for` replaced in this file? Hard-coding URLs could be brittle if things move.

Reply:
I really don't see the point of `url_for`. We don't use it in the `.md` files so if thinks move then we're in trouble already.

We need to remove this or importing this for the ebook doesn't work. Can't remember why so can have a look but, as I say, I don't see the point of this function given we "hardcode" the URLs in the `.md` files.

Reply:
`url_for` is useful so we don't need the templates to care about how the URL are structured by the router in `main.py`. This abstraction gives us flexibility if needed.

I'm not strongly advocating for keeping it, just wanted to hear your reasoning for removing it when I think it provides value.

But as long as you're doing a regex replace of the href values anyway, why does it matter if they're `./` or the full URL?

Reply:
Also we pass in the template name to `url_for` so if we move them about won't we need to change the inputs to the `url_for` function anyway? At which point why no just do a search for `./.*` and fix the "hardcoded" urls?

Reply:
The name passed into `url_for` is the function name of the route defined in `main.py`.

Reply:
OK retested and remember why I did this now. The regex in `main.py` is below:

https://github.com/HTTPArchive/almanac.httparchive.org/blob/3db26fac48360d459dda542191da50173a09db64/src/main.py#L119-L137

so without this change to "hardcode" URLs we won't touch those reference, so that means when the ebook is generated the URL is:
```html
<a href="/en/2019/ecommerce">Ecommerce</a>
```
as opposed to:
```html
<a href="#ecommerce">Ecommerce</a>
```
(which is the internal link to that chapter in the ebook).

So the links don't work within the PDF.

Also I can't change this to:
```
{{ url_for('chapter', year=year, lang=lang, chapter='#ecommerce') }}
```
Or even:
```
{{ url_for('chapter', year=year, lang=lang, chapter='ecommerce',ebook=True) }}
```

Because that will break the real Methodology chapter and I'm reusing this.

I could just regex for this in the get_ebook_methodology function:
```
{{ url_for('chapter', year=year, lang=lang, chapter='#ecommerce') }}
```
and replace it with the hardcoded link in the above Python code, but why not use same hardcoded link for ebook and Methodology page to avoid having to do this? The hardcoding works for both.

So there is a good reason to do this change! 😀

Reply:
My suggestion is to keep the existing `url_for` code throughout the methodology template and update `get_ebook_methodology` to map href URLs like `/en/2019/ecommerce` to `#ecommerce`. (Or maybe using a more unique ID like `#chapter-ecommerce` in case of heading collisions :))

Is that doable?

Reply:
Or...

I can replace this in `main.py`:

```
    methodology_maincontent = re.sub('href="\.\/', 'href="#', methodology_maincontent)
```

to this:

```
    methodology_maincontent = re.sub('href="\/%s\/%s\/' % (lang, year), 'href="#', methodology_maincontent)
    methodology_maincontent = re.sub('href="\.\/', 'href="#', methodology_maincontent)
```

And then I can leave the `url_for` function. Probably no harm to do that anyway, just in case someone does add `url_for` back into the methodology page. Will do that!

And actually to handle other years and languages I should add one more row:

```
    methodology_maincontent = re.sub('href="\/%s\/%s\/' % (lang, year), 'href="#', methodology_maincontent)
    methodology_maincontent = re.sub('href="\/', 'href="https://almanac.httparchive.org/', methodology_maincontent)
    methodology_maincontent = re.sub('href="\.\/', 'href="#', methodology_maincontent)
```

Reply:
> My suggestion is to keep the existing `url_for` code throughout the methodology template and update `get_ebook_methodology` to map href URLs like `/en/2019/ecommerce` to `#ecommerce`. (Or maybe using a more unique ID like `#chapter-ecommerce` in case of heading collisions :))
> 
> Is that doable?

Yup - just figured out how to do it.

And the chapters headers are given links like `#chapter-header` (for example `#performance-first-contentful-paint`) to avoid name clashes. The `#chapter` links are resevered for the chapter names themselves.

Reply:
Perfect, thanks for thinking through this with me!

Reply:
> Perfect, thanks for thinking through this with me!

No, no thank you for the great feedback - keep it coming! This makes the ebook more robust and better.

All the `url_for` have been reverted so going to mark this as resolved. Un-resolve if again if you spot one I missed.